packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+    if (typeof type !== 'string') {
+      throw new InvalidCaveatTypeError(requestedCaveat, origin, target);
+    }
+
+    if (!hasProperty(requestedCaveat, 'value')) {
+      throw new CaveatMissingValueError(requestedCaveat, origin, target);
+    }
+
+    // Assert that the caveat value is valid JSON
+    if (!deepEqual(value, JSON.parse(JSON.stringify(value)))) {
+      throw new CaveatInvalidJsonError(requestedCaveat, origin, target);
+    }
+
+    const caveat = _constructCaveat(type, (value as Json) ?? null);
+    this.validateCaveat(caveat, origin, target);
+    return caveat as ExtractCaveats<ControllerCaveatSpecification>;
+  }
+
+  /**
+   * Validates the specified caveat by ensuring that a specification exists for
+   * its type and invoking the validator function with it, if any.
+   *
+   * @param caveat - The caveat object to validate.
+   * @param origin - The origin associated with the subject of the parent
+   * permission.
+   * @param target - The target name associated with the parent permission.
+   */
+  private validateCaveat(
+    caveat: CaveatConstraint,
+    origin: OriginString,
+    target: string,
+  ): void {
+    const specification = this.getCaveatSpecification(caveat.type);
+
+    if (!specification) {
+      throw new UnrecognizedCaveatTypeError(caveat.type, origin, target);
+    }
+
+    specification.validator?.(caveat, origin, target);
+  }
+
+  /**
+   * Initiates a permission request that requires user approval. This should
+   * always be used to grant additional permissions to a subject, unless user
+   * approval has been obtained through some other means.
+   *
+   * @see {@link ApprovalController} For the user approval logic.
+   * @see {@link PermissionController.acceptPermissionsRequest} For the method
+   * that _accepts_ the request and resolves the user approval promise.
+   * @see {@link PermissionController.rejectPermissionsRequest} For the method
+   * that _rejects_ the request and the user approval promise.
+   * @param origin - The origin of the grantee subject.
+   * @param requestedPermissions - The requested permissions.
+   * @param id - The id of the permissions request.
+   * @param preserveExistingPermissions - Whether to preserve the subject's
+   * existing permissions.
+   * @returns The granted permissions and request metadata.
+   */
+  async requestPermissions(
+    origin: OriginString,
+    requestedPermissions: RequestedPermissions,
+    id: string = nanoid(),
+    preserveExistingPermissions = true,
+  ): Promise<
+    [
+      Record<
+        string,
+        ExtractPermission<
+          ControllerPermissionSpecification,
+          ControllerCaveatSpecification
+        >
+      >,
+      { id: string; origin: OriginString },
+    ]
+  > {
+    this.validateRequestedPermissions(origin, requestedPermissions);
+    const metadata: PermissionsRequest['metadata'] = {
+      id,
+      origin,
+    };
+
+    const permissionsRequest: PermissionsRequest = {
```

Comment:
Nit: these types can be inferred:

```suggestion
    const metadata = {
      id,
      origin,
    };

    const permissionsRequest = {
```